mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -917,11 +916,15 @@ private[spark] object RandomForest extends Logging {
       // being spun up that will definitely do no work.
       val numPartitions = math.min(continuousFeatures.length, input.partitions.length)
 
+      val numInput = input.count()
+      val bcNumInput = input.sparkContext.broadcast(numInput)
+
       input
         .flatMap(point => continuousFeatures.map(idx => (idx, point.features(idx))))
```

Comment:
instead of adding the filter method there, here you can avoid the generation of the record itself if `point.features(idx)` is 0.0